### ISSUE TITLE
장소간 유사도 구하기 (초기)

### ISSUE THREAD

**Issue by subbni** [subbni/lettrip-machine-learning#2]
`import pandas as pd
import numpy as np
from sklearn.metrics.pairwise import cosine_similarity

df = pd.read_csv('final.csv')
df.columns = ['userId','placeName','address','rating']

ratings = df.copy()

ratings_matrix = ratings.pivot_table('rating',index='userId',columns='placeName')
ratings_matrix = ratings_matrix.fillna(0)

ratings_matrix_T = ratings_matrix.transpose()
ratings_matrix_T

item_sim = cosine_similarity(ratings_matrix_T,ratings_matrix_T)

item_sim_df = pd.DataFrame(data=item_sim, index=ratings_matrix.columns,
                          columns=ratings_matrix.columns)
print(item_sim_df.shape)
item_sim_df.head(30)

### 원하는 장소와 유사도가 높은 장소 n개 
item_sim_df["경복궁"].sort_values(ascending=False)[:10]`

**Comment by subbni** [comment]
final.csv 공유 (사용자 28명 ver)

[final.csv](https://github.com/subbni/lettrip-machine-learning/files/12384845/final.csv)

**Comment by subbni** [comment]
### 참고 블로그

- https://ai-with-sudal-ee.tistory.com/6
- https://dacon.io/competitions/official/235946/codeshare/6140
- https://western-sky.tistory.com/58

**Comment by bluemango0312** [comment]
**[서정후]**
[Uploading jeonghu.csv…]()

**Comment by bluemango0312** [comment]
특정 사용자가 방문한 장소들을 기반으로 유사도가 높은 10개의 장소 이름과 유사도를 출력하도록 하였

`#knn 알고리즘 사용/ItemBaseRecommender 클래스 사용

import pandas as pd
import numpy as np
import tensorflow as tf
from tensorflow import keras
from sklearn.metrics.pairwise import cosine_similarity

# 데이터 로드 및 전처리
csv_path = '/content/sample_data/total.csv'
data = pd.read_csv(csv_path, encoding='utf-8')

# Rating 정규화
rating_mean = data['rating'].mean() #평균 계산 
data['rating_normalized'] = data['rating'] - rating_mean

# 고유 사용자 및 아이템 목록 생성
user_ids = data['userId'].unique()
item_ids = data['placeName'].unique()

# 사용자 및 아이템 ID를 매핑하는 딕셔너리 생성
user_id_to_idx = {user_id: idx for idx, user_id in enumerate(user_ids)}
item_id_to_idx = {item_id: idx for idx, item_id in enumerate(item_ids)}

# 아이템-사용자 평점 행렬 생성
num_users = len(user_ids) #사용자 수 계산
num_items = len(item_ids)
rating_matrix = np.zeros((num_items, num_users)) #평점 행렬 0으로 초기화

for _, row in data.iterrows(): #사용자 인덱스랑 아이템 인덱스 찾고 평점을 평균에서 뺀 값을 행렬에 저장
    user_idx = user_id_to_idx[row['userId']]
    item_idx = item_id_to_idx[row['placeName']]
    rating_matrix[item_idx, user_idx] = row['rating_normalized']

# 아이템 간 코사인 유사도 계산
item_similarity = cosine_similarity(rating_matrix)

# 추천 모델 구성 - 케라스 Model 클래스 상속
#ItemBaseRecommender 클래스(인공 신경망)는 item_similarity 행렬을 가중치로 가지며, 유사도를 기반으로 평점을 계산하고 추천 아이템 평점을 반환함.
class ItemBasedRecommender(keras.Model):
    def __init__(self, num_items, item_similarity):
        super(ItemBasedRecommender, self).__init__()
        self.item_similarity = self.add_weight("item_similarity", shape=(num_items, num_items), initializer=tf.keras.initializers.Constant(item_similarity), trainable=False)
        
    def call(self, inputs): #유사도를 기반으로 추천 아이템의 평점 계산
        user_idx, = inputs #모델의 입력(사용자 인덱스 값)
        user_ratings = tf.gather(self.item_similarity, user_idx) #아이템간 유사도 정보 추출
        recommendations = tf.linalg.matmul(self.item_similarity, tf.expand_dims(user_ratings, axis=-1))
        return tf.squeeze(recommendations, axis=-1)

#모델 생성
model = ItemBasedRecommender(num_items, item_similarity) 


# 모델 학습 - 미구현


# 예측 결과 및 유사도 출력
user_id = 10441490830  # 실제 사용자 ID를 입력
test_user_idx = user_id_to_idx.get(user_id, None) #사용자 인덱스 찾기
if test_user_idx is not None: 
    recommendations2 = model([test_user_idx]) #추천 아이템 평점 계산
    sorted_item = np.argsort(recommendations2)[::-1][:10]  # 상위 10개 아이템만 추천/내림차순 정렬

    print(f"Recommended items for user {user_id}:")
    for item_idx in sorted_item:
        item_id = item_ids[item_idx]
        similarity_score = item_similarity[item_idx, test_user_idx]
        print(f"Item: {item_id}, Similarity Score: {similarity_score:.4f}")
else:
    print(f"User {user_id} not found in the data.")`

**<결과 예시>**
![image](https://github.com/subbni/lettrip-machine-learning/assets/96233329/eff31db4-066e-4794-ac3f-84ebf11a61cd)

**Comment by bluemango0312** [comment]
[머신러닝 모델 코드 수정]
- 지역 입력 추가
- 예측 평점 출력 추가
- 리뷰 개수 제한 추가


```
import pandas as pd
import numpy as np
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.metrics import mean_absolute_error, mean_squared_error

# 데이터 로드 및 전처리
csv_path = '/content/sample_data/data_60.csv'
data = pd.read_csv(csv_path, encoding='utf-8')

# Rating 정규화
rating_mean = data['rating'].mean()
data['rating_normalized'] = data['rating'] - rating_mean

user_ids = data['userId'].unique()
item_ids = data['placeName'].unique()

user_id_to_idx = {user_id: idx for idx, user_id in enumerate(user_ids)}
item_id_to_idx = {item_id: idx for idx, item_id in enumerate(item_ids)}

num_users = len(user_ids)
num_items = len(item_ids)
rating_matrix = np.zeros((num_users, num_items))

for _, row in data.iterrows():
    user_idx = user_id_to_idx[row['userId']]
    item_idx = item_id_to_idx[row['placeName']]
    rating_matrix[user_idx, item_idx] = row['rating_normalized']

#코사인 유사도 구하기
user_similarity = cosine_similarity(rating_matrix)

# 사용자로부터 도시명 입력 받기
city_name = input("원하는 도시명을 입력하세요: ")

# CSV 파일의 도시명과 비교하여 일치하는 장소 추출
matching_places = data[data['address'].fillna('').str.contains(city_name)]['placeName'].unique()

# 사용자 간 유사도를 기반으로 추천
user_id = 11318597683
test_user_idx = user_id_to_idx.get(user_id, None)

if test_user_idx is not None:
    reference_users = np.argsort(user_similarity[test_user_idx])[::-1][1:6]
    user_visited_items = np.where(rating_matrix[test_user_idx] > 0)[0]

    recommended_scores = np.zeros(num_items)
    avg_user_rating = rating_matrix[test_user_idx, :].mean()

    for user_idx in reference_users:
        user_visited_items_ref = np.where(rating_matrix[user_idx] > 0)[0]
        non_visited_items = np.setdiff1d(user_visited_items_ref, user_visited_items)
        for item_idx in non_visited_items:
            sim_user_rating = rating_matrix[user_idx, item_idx]
            recommended_scores[item_idx] += user_similarity[test_user_idx, user_idx] * (sim_user_rating - avg_user_rating)

    user_ratings_predicted = np.zeros(num_items)  # 비어있는 예측 평점 행렬 만들기
    for item_idx in range(num_items):
        if item_idx not in user_visited_items:
            similar_items = np.argsort(user_similarity[test_user_idx])[::-1]
            sum_similarity = 0
            sum_weighted_rating = 0
            for sim_item_idx in similar_items:
                if rating_matrix[sim_item_idx, item_idx] > 0:
                    sum_similarity += user_similarity[test_user_idx, sim_item_idx]
                    sum_weighted_rating += user_similarity[test_user_idx, sim_item_idx] * (rating_matrix[sim_item_idx, item_idx] - avg_user_rating)
            if sum_similarity > 0:
                predicted_rating = avg_user_rating + (sum_weighted_rating / sum_similarity)
                user_ratings_predicted[item_idx] = max(1, min(5, predicted_rating))  # 예측 평점을 1부터 5 사이로 제한
                recommended_scores[item_idx] += user_ratings_predicted[item_idx] - avg_user_rating  # 추천 점수를 계산
                recommended_scores[item_idx] = max(0, min(100, (avg_user_rating + recommended_scores[item_idx]) * 20))  # 추천 점수를 0%에서 100%로 변환

    sorted_recommended = np.argsort(recommended_scores)[::-1]
    recommended_places = [(item_ids[idx], recommended_scores[idx]) for idx in sorted_recommended if item_ids[idx] in matching_places][:10]  # 최대 10개의 추천 장소 선택

    print(f"{user_id} 사용자의 {city_name} 지역 추천 장소 리스트")

    # 데이터 출력
    table_data = []
    for place, score in recommended_places:
        user_rating_pred = user_ratings_predicted[item_id_to_idx[place]]
        table_data.append([place, f"{score:.3f}%", user_rating_pred])

    table_columns = ['PlaceName', '추천 점수(%)', '예측 평점']
    table_df = pd.DataFrame(table_data, columns=table_columns)
    display(table_df)

    # 실제 평점과 예측 평점 계산하여 MSE와 MAE 출력
    # 예측 평점 = Σ(유사도 * (아이템의 평점 - 평균 평점)) / Σ|유사도|
    user_ratings_actual = rating_matrix[test_user_idx, :]  # 사용자의 모든 아이템에 대한 평점 데이터 가져오기
    user_ratings_predicted = np.zeros(num_items)  #비어있는 예측 평점 행렬 만들기
    for item_idx in range(num_items):
        if item_idx not in user_visited_items:
            similar_items = np.argsort(user_similarity[test_user_idx])[::-1]
            for sim_item_idx in similar_items:
                if rating_matrix[test_user_idx, sim_item_idx] > 0:
                    user_ratings_predicted[item_idx] += user_similarity[test_user_idx, sim_item_idx] * rating_matrix[test_user_idx, sim_item_idx]
    mae = mean_absolute_error(user_ratings_actual, user_ratings_predicted)
    mse = mean_squared_error(user_ratings_actual, user_ratings_predicted)
    print(f"MAE: {mae:.4f}, MSE: {mse:.4f}")
else:
    print(f"사용자 {user_id} 데이터에 없음")
```
